src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1084,6 +1084,11 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
 
                 if (localSymbol.RefKind != RefKind.None)
                 {
+                    if (IsDirectlyInIterator)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_BadIteratorLocalType, localSymbol.Locations[0]);
```

Comment:
>ERR_BadIteratorLocalType [](start = 50, length = 24)

We used to report this error regardless whether there is an initializer or not.

Reply:
Also, consider bundling this check with the check for ERR_BadAsyncLocalType above. And sharing the code with BindForEachPartsWorker that has similar logic. BTW, why do we need the checks performed in BindForEachPartsWorker? Are we not calling this method for the locals checked there?

---
In reply to: [424592982](https://github.com/dotnet/roslyn/pull/44194#discussion_r424592982) [](ancestors = 424592982)

Reply:
> We used to report this error regardless whether there is an initializer or not.
>
> Also, consider bundling this check with the check for ERR_BadAsyncLocalType above. And sharing the code with BindForEachPartsWorker that has similar logic.

Thanks, will extract out a method which gives diagnostics for 'ref' variables in async or iterator methods.

> BTW, why do we need the checks performed in BindForEachPartsWorker? Are we not calling this method for the locals checked there?

We do not call BindVariableDeclaration for a foreach iteration variable.